### ISSUE TITLE
Pushing from adjusted branch to adjusted branch does not 

### ISSUE THREAD

**Issue by mlell** [datalad/datalad#7141]
# Create upstream repository in unlocked mode
datalad create upstream
cd upstream/
echo test > test1
datalad save -m test1
git annex adjust --unlock

# clone -> downstream repository
cd ..
datalad clone upstream downstream
cd downstream/
echo test2> test1
datalad save -m test2
datalad push

# The adjusted branch of upstream not updated

<details>

**Comment by mlell** [comment]
Performing `git annex sync` in the `upstream` repository resolves the problem:

```sh
.../upstream$ git annex sync
commit
On branch adjusted/master(unlocked)
nothing to commit, working tree clean
ok
merge synced/master (Merging into master...)
Updating a09b4d5..a6cad42
Fast-forward
 test1 | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
(Merging into adjusted branch...)
Updating 7bb5868..54232b5
Fast-forward
 test1 | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)
ok

.../upstream$ git log --oneline --graph
* dca42ae (HEAD -> adjusted/master(unlocked)) git-annex adjusted branch
* a6cad42 (synced/master, master, refs/basis/adjusted/master(unlocked)) test2
* a09b4d5 test1
* 09d18ef [DATALAD] new dataset
```

However, `git annex sync` does a whole lot of stuff which I do not completely understand, for example if the upstream repository has another remote itself...

**Comment by bpoldrack** [comment]
> git annex sync does a whole lot of stuff which I do not completely understand

Note, that in this situation you only need a "local sync", avoiding any other operation `annex sync` would possibly do, by calling `git annex sync --no-push --no-pull --no-commit --no-resolvemerge --no-content`.

> Is there a possibilty that datalad does something like that automatically to the upstream repository?

Trying to do something on the remote end sounds like a troublesome approach to me. I suppose the "clean" solution is to set git's post-update hook in upstream accordingly (executing above mentioned call).
What we could do, I think, is providing a config procedure to set up the upstream repo that way. This would then be executed via `datalad create -c <the new procedure> ...`.
Would that be a workable solution for you, @mlell?

WDYT, @datalad/developers?

**Comment by bpoldrack** [comment]
> git annex sync does a whole lot of stuff which I do not completely understand

Note, that in this situation you only need a "local sync", avoiding any other operation `annex sync` would possibly do, by calling `git annex sync --no-push --no-pull --no-commit --no-resolvemerge --no-content`.

> Is there a possibilty that datalad does something like that automatically to the upstream repository?

Trying to do something on the remote end sounds like a troublesome approach to me. I suppose the "clean" solution is to set git's post-update hook in upstream accordingly (executing above mentioned call).
What we could do, I think, is providing a config procedure to set up the upstream repo that way. This would then be executed via `datalad create -c <the new procedure> ...`.
Would that be a workable solution for you, @mlell?

WDYT, @datalad/developers?

**Comment by mlell** [comment]
> Trying to do something on the remote end sounds like a troublesome approach to me. I suppose the "clean" solution is 
> to set git's post-update hook in upstream accordingly (executing above mentioned call).
So do I understand correctly that your concern here is that you want the remote machine (that has the upstream repo) to do the updating, hence the post-update hook? That sounds reasonable to me.

> What we could do, I think, is providing a config procedure to set up the upstream repo that way. This would then be 
> executed via datalad create -c <the new procedure> ....
> Would that be a workable solution for you, @mlell?
That solution would, at least for me, not be what I would expect following the "law of least surprise", because it differs from what git does. Git rejects a push to a checked-out branch, except if `git config receive.denyCurrentBranch=updateInstead` is set. Therefore, in datalad, a `datalad push` fails if upstream has the pushed-to branch checked out, which feels consistent to me (even though novice users might profit from a more detailed error message than: "`publish(error): . (dataset) [refs/heads/master->origin:refs/heads/master [remote rejected] (branch is currently checked out)]`". For adjusted branches, what would feel consistent to me is that the push is rejected if either the branch (status quo) or its adjusted branch (currently not so) are checked out.

Then, the user can choose to set `git config receive.denyCurrentBranch=updateInstead` (either directly or following a datalad help message or a `create -c` procedure like you suggested (can that be run later as well?)) so that both the master branch and the adjusted branch are updated on the remote.

**Comment by mlell** [comment]
Also, it seems that `git annex sync --no-push --no-pull --no-commit --no-resolvemerge --no-content` does not resolve the upstream state for me:

```sh
.../upstream [adjusted/master(unlocked) L|✔]
$ git annex sync --no-push --no-pull --no-commit --no-resolvemerge --no-content

.../upstream [adjusted/master(unlocked) L|✔]
$ git log --graph --oneline HEAD master
* 064054e (HEAD -> adjusted/master(unlocked)) git-annex adjusted branch
* 2d8c26b (synced/master, master, refs/basis/adjusted/master(unlocked)) test1
* cfe7549 [DATALAD] new dataset

.../upstream [adjusted/master(unlocked) L|✔]
$ cd ../downstream/

.../downstream [adjusted/master(unlocked) ↓·1↑·2|✔]
$ git log --graph --oneline HEAD master
* 8025be0 (HEAD -> adjusted/master(unlocked)) git-annex adjusted branch
* 54574b3 (master, refs/basis/adjusted/master(unlocked)) test2
* 2d8c26b (origin/master) test1
* cfe7549 [DATALAD] new dataset

```

**Comment by bpoldrack** [comment]
> So do I understand correctly that your concern here is that you want the remote machine [...] to do the updating

Yes. Whether something runs on my repo should at least technically be under my control, not the control of someone who pushes into it.

> procedure like you suggested (can that be run later as well?)

Yes, could be run anytime via `datalad run-procedure`.

> ... For adjusted branches, therefore the behaviour should be the same, I think.

I do understand that point of view. I guess it's clear that technically it's not what is happening. You don't really push an adjusted branch to a checked out adjusted branch. But certainly feels like this is what's happening.

Maybe I forgot something with respect to adjusted branches and annex, but to some extend this:
> because it differs from what git does

must remain the case, since git does not have any idea of such "coupled" branches. There's no equivalent in git.

In any case, need to think that through a bit.

**Comment by bpoldrack** [comment]
> does not resolve the upstream state for me, but reverses the push:

Ok, that is confusing me right now. Something is messed up, either with the push itself or my brain. Thanks for the details, will dig into this once the fog in my head cleared ;-)
